dlt/sources/rest_api/typing.py

```diff
@@ -0,0 +1,279 @@
+
+    base_page: Optional[int]
+    page_param: Optional[str]
+    total_path: Optional[jsonpath.TJsonPath]
+    maximum_page: Optional[int]
+
+
+class OffsetPaginatorConfig(PaginatorTypeConfig, total=False):
+    """A paginator that uses offset-based pagination strategy."""
+
+    limit: int
+    offset: Optional[int]
+    offset_param: Optional[str]
+    limit_param: Optional[str]
+    total_path: Optional[jsonpath.TJsonPath]
+    maximum_offset: Optional[int]
+
+
+class HeaderLinkPaginatorConfig(PaginatorTypeConfig, total=False):
+    """A paginator that uses the 'Link' header in HTTP responses
+    for pagination."""
+
+    links_next_key: Optional[str]
+
+
+class JSONLinkPaginatorConfig(PaginatorTypeConfig, total=False):
+    """Locates the next page URL within the JSON response body. The key
+    containing the URL can be specified using a JSON path."""
+
+    next_url_path: Optional[jsonpath.TJsonPath]
+
+
+class JSONResponseCursorPaginatorConfig(PaginatorTypeConfig, total=False):
+    """Uses a cursor parameter for pagination, with the cursor value found in
+    the JSON response body."""
+
+    cursor_path: Optional[jsonpath.TJsonPath]
+    cursor_param: Optional[str]
+
+
+PaginatorConfig = Union[
+    PaginatorType,
+    PageNumberPaginatorConfig,
+    OffsetPaginatorConfig,
+    HeaderLinkPaginatorConfig,
+    JSONLinkPaginatorConfig,
+    JSONResponseCursorPaginatorConfig,
+    BasePaginator,
+    SinglePagePaginator,
+    HeaderLinkPaginator,
+    JSONLinkPaginator,
+    JSONResponseCursorPaginator,
+    OffsetPaginator,
+    PageNumberPaginator,
+]
+
+
+AuthType = Literal["bearer", "api_key", "http_basic"]
+
+
+class AuthTypeConfig(TypedDict, total=True):
+    type: AuthType  # noqa
+
+
+class BearerTokenAuthConfig(TypedDict, total=False):
+    """Uses `token` for Bearer authentication in "Authorization" header."""
+
+    # we allow for a shorthand form of bearer auth, without a type
+    type: Optional[AuthType]  # noqa
+    token: str
+
+
+class ApiKeyAuthConfig(AuthTypeConfig, total=False):
+    """Uses provided `api_key` to create authorization data in the specified `location` (query, param, header, cookie) under specified `name`"""
+
+    name: Optional[str]
+    api_key: str
+    location: Optional[TApiKeyLocation]
+
+
+class HttpBasicAuthConfig(AuthTypeConfig, total=True):
+    """Uses HTTP basic authentication"""
+
+    username: str
+    password: str
+
+
+# TODO: add later
+# class OAuthJWTAuthConfig(AuthTypeConfig, total=True):
+
+
+AuthConfig = Union[
+    AuthConfigBase,
+    AuthType,
+    BearerTokenAuthConfig,
+    ApiKeyAuthConfig,
+    HttpBasicAuthConfig,
+    BearerTokenAuth,
+    APIKeyAuth,
+    HttpBasicAuth,
+]
+
+
+class ClientConfig(TypedDict, total=False):
+    base_url: str
+    headers: Optional[Dict[str, str]]
+    auth: Optional[AuthConfig]
+    paginator: Optional[PaginatorConfig]
+
+
+class IncrementalArgs(TypedDict, total=False):
+    cursor_path: str
+    initial_value: Optional[str]
```

Comment:
Using the generic `TCursorValue` instead of `str` we get the following errors:

```
E                       dlt.common.exceptions.DictValidationException: In path .: field 'resources[0]' expects the following types: str, EndpointResource. Provided value {'name': 'posts', 'endpoint': {'path': 'posts', 'incremental': {'start_param': 'since', 'end_param': 'until', 'cursor_path': 'updated_at', 'initial_value': '1', 'end_value': '86401', 'convert': <function test_posts_with_inremental_date_conversion.<locals>.<lambda> at 0x15abca980>}}} with type 'dict' is invalid with the following errors:
E                       For EndpointResource: In path ./resources[0]: field 'endpoint' expects the following types: str, Endpoint. Provided value {'path': 'posts', 'incremental': {'start_param': 'since', 'end_param': 'until', 'cursor_path': 'updated_at', 'initial_value': '1', 'end_value': '86401', 'convert': <function test_posts_with_inremental_date_conversion.<locals>.<lambda> at 0x15abca980>}} with type 'dict' is invalid with the following errors:
E                       For Endpoint: In path ./resources[0]/endpoint/incremental: field 'cursor_path' has expected type 'TCursorValue' which lacks validator
E                       For str: In path ./resources[0]: field 'endpoint' with value {'path': 'posts', 'incremental': {'start_param': 'since', 'end_param': 'until', 'cursor_path': 'updated_at', 'initial_value': '1', 'end_value': '86401', 'convert': <function test_posts_with_inremental_date_conversion.<locals>.<lambda> at 0x15abca980>}} has invalid type 'dict' while 'str' is expected
```

Reply:
We decided to leave out support for non string configs in this PR and I create a follow-up issue: https://github.com/dlt-hub/dlt/issues/1757

It seems that our implementation already supports int etc. but we'd need to implement the dict validation.